piedpiper/gman/client.py

```diff
@@ -0,0 +1,255 @@
+
+
+def get_thread_id_tasks(thread_id=None, gman_url=None, query_filter=None):
+    """
+    Get a list of tasks associated with the given thread_id
+    :param thread_id: The thread_id to query with as a sring
+    :param gman_url: The GMan endpoint as a string
+    :param query_filter: lambda expression
+    :return: List of tasks associated with thread_id
+    """
+    try:
+        r = requests.get(f"{gman_url}/thread/{thread_id}")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def get_thread_id_events(thread_id=None, gman_url=None, query_filter=None):
+    """
+    Get list of all events for a given thread_id, optionally filtered by
+    a lambda expression.
+    :param thread_id: Thread ID to query for as a string.
+    :param gman_url:  GMan endpoint as a string.
+    :param query_filter: Lambda expression
+    :return: List of task events
+    """
+    try:
+        r = requests.get(f"{gman_url}/thread/{thread_id}/events")
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        raise requests.exceptions.HTTPError(
+            f"Gman returned with a bad status code. \n\n{e}"
+        )
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            f"Failed to get tasks for thread_id {thread_id}. \n\n{e}"
+        )
+    else:
+        if query_filter:
+            return list(filter(query_filter, r.json()))
+        else:
+            return r.json()
+
+
+def wait_for_thread_id_complete(thread_id=None, gman_url=None, retry_max=10):
+    """
+    Wait for all tasks under a given thread_id to return with a status of complete,
+    up to the retry_max.
+    :param thread_id: thread_id as a string to search for
+    :param gman_url: GMan endpoint as a string
+    :param retry_max: Number of times to retry as an integer
+    :return: True or exception
+    """
+    retries = 0
+    while retries < retry_max:
+        try:
+            log.debug(f"Checking status of thread {thread_id}")
+            r = requests.head(f"{gman_url}/thread/{thread_id}")
+            r.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            raise requests.exceptions.HTTPError(
+                f"GMan returned with a bad status code. \n\n{e}"
+            )
+        except requests.exceptions.RequestException as e:
+            raise requests.exceptions.HTTPError(
+                f"Failed to check status of task. \n\n{e}"
+            )
+        else:
```

Comment:
Same as above, try: except: else: is used to handle stuff where an exception is not thrown, but the general paradigm is try: except: finally: else can almost always be included in the try: section with the one exception if you want to raise the same type of error you are attempting to catch ... then an else might allow easier logic for those cases.